### PR TITLE
update to I.4: Make interfaces precisely and strongly typed

### PR DESCRIPTION
**Pull request by ChrisGuzak**
This relates to issue #1288

### CODE REVIEW

**Comment by ChrisGuzak** [comment]
This relates to #1288

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -1391,18 +1391,23 @@ Very hard in general.
 
 ##### Reason
 
-Types are the simplest and best documentation, have well-defined meaning, and are guaranteed to be checked at compile time.
+Types are the simplest and best documentation, improving legibility due to their well-defined meaning, and are checked at compile time.
```
Comment (blakehawkins) [inline CppCoreGuidelines.md]: /improving/improve/

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -1391,18 +1391,23 @@ Very hard in general.
...
 Also, precisely typed code is often optimized better.
 
 ##### Example, don't
 
 Consider:
 
-    void pass(void* data);    // void* is suspicious
+    void pass(void* data);    // weak and under qualified typing is suspicious
```
Comment (blakehawkins) [inline CppCoreGuidelines.md]: /typing/type void*/

File: CppCoreGuidelines.md [hunk 3]
```diff
@@ -1391,18 +1391,23 @@ Very hard in general.
...
 
-Now the callee must cast the data pointer (back) to a correct type to use it. That is error-prone and often verbose.
-Avoid `void*`, especially in interfaces.
-Consider using a `variant` or a pointer to base instead.
+Callers are unsure what types are allowed and if the data may
+be mutated as `const` is not specified. Note all pointer types
+ impliclty convert to void* so it is easy for callers to provide this value.
```
Comment (blakehawkins) [inline CppCoreGuidelines.md]: Weird line breaks, missing commas

File: CppCoreGuidelines.md [hunk 4]
```diff
@@ -1411,12 +1416,12 @@ For generic code these `T`s can be general or concept constrained template param
 
 Consider:
 
-    void draw_rect(int, int, int, int);   // great opportunities for mistakes
+    draw_rect(100, 200, 100, 500); // what do the numbers specify?
 
-    draw_rect(p.x, p.y, 10, 20);          // what does 10, 20 mean?
+    draw_rect(p.x, p.y, 10, 20); // what units are 10 and 20 in?
+
+It is clear that the caller is describing a rect, but it is unclear what parts they relate to. Also, an `int` can carry values of many units, so we must guess about their meaning.
```
Comment (blakehawkins) [inline CppCoreGuidelines.md]: /guess about/guess/

**Comment by cubbimew** [comment]
Line 1405: impliclty -> simplicity
Line 1410: undescribable -> indescribable
Line 1423: rect -> cert  (rectangle could work here)

**Pull request merged**